File: 본문/python/Code_4_05_2.py
import sys

# 깊이 우선 탐색을 하는 함수
def dfs(pos, G, visited):
	visited[pos] = True
	for i in G[pos]:
		if visited[i] == False:
			dfs(i, G, visited)

# 재귀 호출 상한을 120000으로 설정
sys.setrecursionlimit(120000)

# 입력
N, M = map(int, input().split())
A = [ None ] * M
B = [ None ] * M
for i in range(M):
	A[i], B[i] = map(int, input().split())

# 인접리스트 작성
G = [ list() for i in range(N + 1) ]
for i in range(M):
	G[A[i]].append(B[i])
	G[B[i]].append(A[i])

# 깊이 우선 탐색
visited = [ False ] * (N + 1)
dfs(1, G, visited)

# 연결 판정(answer = true일 때 연결)
answer = True
for i in range(1, N + 1):
	if visited[i] == False:
		answer = False
if answer == True:
	print("The graph is connected.")
else:
	print("The graph is not connected.")

File: 본문/cpp/Code_4_05_2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N, M, A[100009], B[100009];
vector<int> G[100009];
bool visited[100009]; // visited[pos]=false일 때 정점x가 흰색, true일 때 회색

void dfs(int pos) {
	visited[pos] = true;
	// for (int i : G[pos])을 "범위기반 for 반복문"이라고 부릅니다.
	for (int i : G[pos]) {
		if (visited[i] == false) dfs(i);
	}
}

int main() {
	// 입력
	cin >> N >> M;
	for (int i = 1; i <= M; i++) {
		cin >> A[i] >> B[i];
		G[A[i]].push_back(B[i]);
		G[B[i]].push_back(A[i]);
	}

	// 깊이 우선 탐색
	dfs(1);

	// 연결되어 있는지 판정(Answer=true라면 연결됨)
	bool Answer = true;
	for (int i = 1; i <= N; i++) {
		if (visited[i] == false) Answer = false;
	}
	if (Answer == true) cout << "The graph is connected." << endl;
	else cout << "The graph is not connected." << endl;
	return 0;
}
